### DISCUSSION TITLE
Best Practice for Upgrading from 0.3.0 to 0.4.5

### DISCUSSION THREAD

**Discussion by Korbi-code** [beeware/toga#2731] [Q&A]
Hello,

It's been quite a while since I started building my first app with Toga. I started with version 0.3.0 and built a decent application that has been running for a year without any issues.

Now, I want to implement new features. Additionally, I thought about the idea of upgrading to the latest version of Toga, as many bug fixes and features have been implemented since then.

Has anyone had any experience upgrading the version? Is there a best practice for doing this?

Many thanks in advance.

**Comment by freakboy3742** [comment]
I don't know that I have any specific advice other than [read the release notes](https://github.com/beeware/toga/releases), and work incrementally. When there are known backwards incompatibilities, we flag those in a "Backwards Incompatible Changes" section; and where possible, we add runtime warnings about APIs that change, and provide a compatibility path for the "old" APIs. If you upgrade one release at a time, and test after each upgrade, you should get exposed to as many of those warnings as possible.

The biggest jump will be between v0.3.1 and v0.4.0. That was a release that was the result of a major API review. As a result, the backwards incompatibilities section for that release is quite large, and there are a lot of small changes that might require a response in your app code. How complex this process is depends very much on what your app is doing, and the extent to which your app's code was relying on undocumented or untested behaviours; but once you're at the 0.4.0 release, getting from 0.4.0 to 0.4.5 *should* be fairly smooth.

**Comment by l-o-u-i-s-g** [comment]
I create a new venv  for  major upgrades.  Create a new project with same name in new venv  and copy source from old to new  and go from there..  so you still have your old venv  if you need quick fixes   while you are still porting  to the new release ..
